src/notebooks/flatten_and_explode.ipynb

```diff
@@ -0,0 +1,524 @@
+    "        # Grab columns that are already flat and move to final df\n",
+    "        parents, df = stack.pop()\n",
+    "        flat_cols = [col(\".\".join(parents + (c[0],))).alias(\"_\".join(parents + (c[0],))) for c in df.dtypes if c[1][:6] != \"struct\"]\n",
+    "        columns.extend(flat_cols)\n",
+    "        \n",
+    "        # Recursively collect columns within each struct, searching until\n",
+    "        # the projection is flat\n",
+    "        nested_cols = [c[0] for c in df.dtypes if c[1][:6] == \"struct\" ]    \n",
+    "        for nc in nested_cols:\n",
+    "            projected_df = df.select(nc + \".*\")\n",
+    "            stack.append((parents + (nc,), projected_df))\n",
+    "            \n",
+    "    return nested_df.select(columns)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 0,
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "inputWidgets": {},
+     "nuid": "4b79cb27-08b1-4736-b4ff-f1ad75fefa21",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/html": [
+       "<style scoped>\n",
+       "  .ansiout {\n",
+       "    display: block;\n",
+       "    unicode-bidi: embed;\n",
+       "    white-space: pre-wrap;\n",
+       "    word-wrap: break-word;\n",
+       "    word-break: break-all;\n",
+       "    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n",
+       "    font-size: 13px;\n",
+       "    color: #555;\n",
+       "    margin-left: 4px;\n",
+       "    line-height: 19px;\n",
+       "  }\n",
+       "</style>\n",
+       "<div class=\"ansiout\"></div>"
+      ]
+     },
+     "metadata": {
+      "application/vnd.databricks.v1+output": {
+       "addedWidgets": {},
+       "arguments": {},
+       "data": "<div class=\"ansiout\"></div>",
+       "datasetInfos": [],
+       "metadata": {},
+       "removedWidgets": [],
+       "type": "html"
+      }
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "'''\n",
+    "Transforms a spark dataframe into an \"exploded\" data frame. To explode a data frame \n",
+    "means to add a new row containing identical identifying information to a base row \n",
+    "for each array-element encountered in columns whose fields are structs. The result\n",
+    "is returned as a dataframe.\n",
+    "'''\n",
+    "def explode_arrays(df):\n",
+    "    flat_cols = [field.name for field in df.schema.fields if type(field.dataType) != ArrayType]\n",
+    "    struct_cols = [field.name for field in df.schema.fields if type(field.dataType) == ArrayType]\n",
+    "    \n",
+    "    # Add new rows for each array element, then select and join with columns\n",
+    "    # that are already flat\n",
+    "    exploded_df = df.withColumn('vals', explode_outer(arrays_zip(*struct_cols))) \\\n",
+    "           .select(*flat_cols,'vals.*') \\\n",
+    "           .fillna('', subset=struct_cols)\n",
+    "    return exploded_df"
```

Comment:
nit: We can make this section of code a little cleaner using the fact that Python considers everything inside of parentheses to be a single line.

```
exploded_df = (
    df.withColumn('vals', explode_outer(arrays_zip(*struct_cols)))
    .select(*flat_cols, 'vals.*')
    .fillna('', subset=struct_cols)
)
```
I think this makes the code cleaner, easier to read, and less prone to errors resulting from for getting the `\` on a line. But there's nothing wrong with how you've implemented it, so go with what you like.